src/mol-model-props/computed/representations/interactions-intra-unit-cylinder.ts

```diff
@@ -39,22 +41,64 @@ async function createIntraUnitInteractionsCylinderMesh(ctx: VisualContext, unit:
     const contacts = interactions.unitsContacts.get(unit.id);
 
     const { x, y, z, members, offsets } = features;
-    const { edgeCount, a, b, edgeProps: { flag } } = contacts;
-    const { sizeFactor, parentDisplay } = props;
+    const { edgeCount, a, b, edgeProps: { flag, type } } = contacts;
+    const { sizeFactor, ignoreHydrogens, ignoreHydrogensVariant, parentDisplay } = props;
 
     if (!edgeCount) return Mesh.createEmpty(mesh);
 
+    const pos = unit.conformation.invariantPosition;
+    const { elements } = unit;
+    const p = Vec3();
+    const pA = Vec3();
+    const pB = Vec3();
+
     const builderProps = {
         linkCount: edgeCount * 2,
         position: (posA: Vec3, posB: Vec3, edgeIndex: number) => {
-            Vec3.set(posA, x[a[edgeIndex]], y[a[edgeIndex]], z[a[edgeIndex]]);
-            Vec3.set(posB, x[b[edgeIndex]], y[b[edgeIndex]], z[b[edgeIndex]]);
+            const t = type[edgeIndex];
+            if ((!ignoreHydrogens || ignoreHydrogensVariant !== 'all') && (
+                t === InteractionType.HydrogenBond || t === InteractionType.WeakHydrogenBond)
+            ) {
+                const idxA = members[offsets[a[edgeIndex]]];
+                const idxB = members[offsets[b[edgeIndex]]];
+                pos(elements[idxA], pA);
+                pos(elements[idxB], pB);
+                let minDistA = Vec3.distance(pA, pB);
+                let minDistB = minDistA;
+                Vec3.copy(posA, pA);
+                Vec3.copy(posB, pB);
+
+                eachIntraBondedAtom(unit, idxA, (_, idx) => {
+                    if (isHydrogen(structure, unit, elements[idx], 'polar')) {
+                        pos(elements[idx], p);
+                        const dist = Vec3.distance(p, pB);
+                        if (dist < minDistA) {
+                            minDistA = dist;
+                            Vec3.copy(posA, p);
+                        }
+                    }
+                });
+
+                eachIntraBondedAtom(unit, idxB, (_, idx) => {
```

Comment:
Creating two closures inside a loop isn't exactly cheap. Tho I guess it shouldn't be a problem in this particular case.

If you feel like it in the future, there is a way to avoid creating the closure with something like this:

```ts
function eachIntraBondedAtom(unit, idx, cb, thisArg?: State) {
   for (const b of ...) {
     cb.call(thisArg, x, y);
   }
}
```

and then the callback can be a top-level function like this:

```ts
function eachBond(this: State, x, y) {
  this.whatever = x;
}
```

Reply:
good point, opened #813